# NumPy Complete Revision Notebook

A structured, example-driven guide to NumPy — covering array creation, attributes, operations, indexing, reshaping, stacking, copies vs views, linear algebra, masked arrays, and more.

## Table of Contents
1. Installation & Setup
2. Introduction to NumPy
3. Creating Arrays (1D / 2D / 3D)
4. Array Attributes
5. Array Creation Functions (zeros, ones, full, empty)
6. `arange()` and `linspace()`
7. Basic Arithmetic Operations
8. Random Number Generation
9. Universal Functions (ufuncs) & Aggregations
10. Indexing, Slicing & Iterating
11. `flatten()` vs `ravel()`
12. Reshaping Arrays
13. Stacking Arrays
14. Splitting Arrays
15. Copies vs Views vs Assignment
16. Sorting Arrays
17. Concatenation
18. Boolean Indexing & Masking
19. Broadcasting
20. Axis-wise Aggregation
21. Unique Elements
22. Transpose
23. Flipping Arrays
24. Using `help()` for Documentation
25. Linear Algebra (Inverse & Determinant)
26. Masked Arrays
27. Interview Questions — Quick Revision (Appendix)

---


## 1. Installation & Setup
Install NumPy using pip (skip this cell if it's already installed in your environment).

In [16]:
#pip install numpy

In [17]:
import numpy as np
print("Numpy Imported Successfully")

Numpy Imported Successfully


## 2. Introduction to NumPy

# NumPy

**NumPy (Numerical Python)** is the fundamental package for **scientific computing in Python**. It is a Python library that provides a powerful **multidimensional array object (`ndarray`)**, various derived objects (such as **masked arrays** and **matrices**), and a wide range of routines for performing fast numerical operations.

## Features of NumPy

NumPy provides functions for:

- Mathematical operations
- Logical operations
- Shape manipulation
- Sorting and searching
- Input and Output (I/O)
- Discrete Fourier Transforms (FFT)
- Basic Linear Algebra
- Statistical operations
- Random number generation and simulation

## Characteristics (Restrictions) of NumPy Arrays

NumPy arrays have the following restrictions:

- All elements in an array must be of the **same data type** (homogeneous).
- Once an array is created, its **size cannot be changed**.
- Arrays must have a **rectangular shape** (not jagged).  
  For example, every row in a two-dimensional array must have the same number of columns.

## Advantages of These Restrictions

Because of these restrictions, NumPy arrays are:

- Faster than Python lists.
- More memory efficient.
- Optimized for numerical and scientific computations.
- Convenient for handling large datasets.

## Performance and Speed

NumPy achieves high performance through:

- **Vectorized Operations:** Performs operations on entire arrays without explicit Python loops, making computations significantly faster.
- **Implementation in C:** Many of NumPy's core functions are written in **C**, providing optimized execution.
- **Contiguous Memory Allocation:** Stores array elements in contiguous blocks of memory, improving CPU cache utilization and reducing processing overhead.

## NumPy's Main Object (`ndarray`)

The primary object in NumPy is the **homogeneous multidimensional array**, called **`ndarray`**.

An `ndarray`:

- Is a table of elements (usually numbers).
- Stores elements of the **same data type**.
- Is indexed using a **tuple of non-negative integers**.
- Can represent **1D, 2D, 3D, or higher-dimensional** data.



## 3. Creating Arrays (1D / 2D / 3D)
Arrays are created with `np.array()` from Python lists/sequences, or generated directly with functions like `np.arange()`. Below we create a 1D array, then a 2D array using `arange().reshape()`.

In [18]:
a=np.array([1,2,3,4,5])  # Initializing a 1D array
print(a)  # Printing the array

[1 2 3 4 5]


In [19]:
a=np.arange(15).reshape(3,5)  # Initializing a 2D array
print(a)  # Printing the array
print(type(a))  # Printing the type of the array

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]
<class 'numpy.ndarray'>


## 4. Array Attributes
Every ndarray exposes metadata about itself: number of dimensions (`ndim`), `shape`, total element count (`size`), element `dtype`, bytes per element (`itemsize`), total bytes (`nbytes`), the raw memory buffer (`data`), and `strides` — how many bytes to skip to move to the next element along each axis.

**Important Attributes**

In [20]:
print(a.ndim)  # Printing the number of dimensions
print(a.shape)  # Printing the shape of the array
print(a.size)  # Printing the total number of elements in the array
print(a.dtype)  # Printing the data type of the array elements
print(a.itemsize)  # Printing the size of each element in bytes
print(a.nbytes)  # Printing the total size of the array in bytes
print(a.data)  # Printing the buffer(memory addr) containing the actual elements of the array

2
(3, 5)
15
int64
8
120


In [21]:
import numpy as np

x = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.int16)
print(x.itemsize)
print(x.strides)  # Output: (6, 2) # how many bytes one must move in memory to go to the next element along each dimension of the array

2
(6, 2)


### A 3D array example
Higher-dimensional arrays behave the same way — `ndim`, `shape`, etc. simply grow to describe the extra axis.

In [22]:
a=np.arange(24).reshape(2,3,4)  # Initializing a 3D array
print(a)  # Printing the array

[[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]


In [23]:
print(a.ndim)  # Printing the number of dimensions
print(a.shape)  # Printing the shape of the array
print(a.size)  # Printing the total number of elements in the array
print(a.dtype)  # Printing the data type of the array elements
print(a.itemsize)  # Printing the size of each element in bytes
print(a.nbytes)  # Printing the total size of the array in bytes
print(a.data)  # Printing the buffer(memory addr) containing the actual elements of the array

3
(2, 3, 4)
24
int64
8
192


In [24]:
c = np.array([[1, 2], [3, 4]], dtype=np.complex128)
c

array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]])

## 5. Array Creation Functions (Placeholder Arrays)
When the actual data isn't known yet, NumPy provides placeholder-array constructors: `zeros()`, `ones()`, `full()`, `empty()` — plus their `*_like()` variants which copy the shape/dtype of an existing array.

to create arrays with initial placeholder content. These minimize the necessity of growing arrays, an expensive operation.

In [25]:
arr_zeros=np.zeros((2,5))
print(arr_zeros)  # Printing the array of zeros
print(arr_zeros.dtype)

[[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]
float64


In [26]:
arr_zeros=np.zeros((2,5),dtype=np.int32)
print(arr_zeros)  # Printing the array of zeros
print(arr_zeros.dtype)

[[0 0 0 0 0]
 [0 0 0 0 0]]
int32


In [27]:
np.zeros_like(arr_zeros) #zeros_like(array) # Creating an array of zeros with the same shape and type as the given array

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]], dtype=int32)

In [28]:
arr_ones=np.ones((2,5))
print(arr_ones)  # Printing the array of ones
print(arr_ones.dtype)

[[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]]
float64


In [29]:
arr_empty=np.empty((2,5)) # Creating an empty array of shape (2,5) with random values
print(arr_empty)  # Printing the empty array

[[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]]


In [30]:
np.empty_like(arr_empty) #empty_like(array) # Creating an empty array with the same shape and type as the given array

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [31]:
np.full_like(arr_empty, 5) #full_like(array, value) # Creating an array with the same shape and type as the given array, filled with the specified value

array([[5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.]])

In [32]:
np.full((2,5), 7) # Creating an array of shape (2,5) filled with the value 7

array([[7, 7, 7, 7, 7],
       [7, 7, 7, 7, 7]])

In [33]:
np.empty((4,5))

array([[4.67296746e-307, 1.69121096e-306, 9.34601642e-307,
        1.89146896e-307, 7.56571288e-307],
       [4.00544531e-307, 1.11261095e-306, 3.56043054e-307,
        1.37961641e-306, 2.22518251e-306],
       [1.33511969e-306, 6.23037657e-307, 9.34609790e-307,
        8.45593934e-307, 9.34600963e-307],
       [1.42418987e-306, 1.60219035e-306, 6.23057349e-307,
        1.69118923e-306, 1.24611470e-306]])

## 6. `arange()` and `linspace()`
`np.arange(start, stop, step)` works like Python's `range()` but returns an array — with float steps it can be hard to predict exactly how many elements you get. `np.linspace(start, stop, num)` solves this by letting you specify the exact **number of evenly spaced elements** instead of a step size.

In [34]:
np.arange(10,30,5)  # Creating an array with values from 10 to 30 with a step of 5

array([10, 15, 20, 25])

In [35]:
np.arange(0,3,0.33)  # Creating an array with values from 0 to 2 with a step of 0.3

array([0.  , 0.33, 0.66, 0.99, 1.32, 1.65, 1.98, 2.31, 2.64, 2.97])

problem with arange with float value is not possible to predict the no of elements.

In [36]:
## sol : linspace() function is used to create an array of evenly spaced values over a specified range. It is particularly useful when you want to generate a specific number of points between two values, including the endpoints.

In [37]:
## we specify no of elements
np.linspace(0,2,9)  # (start,end ,no of elements)  # Creating an array with 9 evenly spaced values between 0 and 2

array([0.  , 0.25, 0.5 , 0.75, 1.  , 1.25, 1.5 , 1.75, 2.  ])

## 7. Basic Arithmetic Operations
Arithmetic operators (`+ - * /`) on arrays apply **element-wise**. Matrix multiplication (dot product) uses `.dot()` or the `@` operator, not `*`.

***Basic Operations***

In [38]:
a=np.array([10,20,30,40])
b=np.array([1,2,3,4])
print(a-b)
print(a+b)
print(a*b)
## arithmetic operartions on arrays are performed element-wise, meaning that the operation is applied to each corresponding pair of elements in the arrays.

[ 9 18 27 36]
[11 22 33 44]
[ 10  40  90 160]


In [39]:
a=np.array([[1,2],[3,4]])
b=np.array([[5,6],[7,8]])
print(a.dot(b))  # Performing matrix multiplication using the dot() function
print(a*b)  # Performing element-wise multiplication using the * operator

[[19 22]
 [43 50]]
[[ 5 12]
 [21 32]]


## 8. Random Number Generation
Modern NumPy recommends `np.random.default_rng(seed)` (a Generator object) over the older global `np.random.rand()` / `np.random.randint()` functions, since a Generator gives reproducible, independent streams of random numbers.

In [40]:
rg=np.random.default_rng(1)  # Creating a random number generator with a seed of 1
print(rg)  # Printing the random number generator object

Generator(PCG64)


In [41]:
a=np.ones((2,3))  # Creating a 2D array of ones with shape (2,3)
a*=3  # Multiplying each element of array 'a' by 3 # inplace operation
print(a)  # Printing the modified array 'a'

[[3. 3. 3.]
 [3. 3. 3.]]


In [42]:
b=rg.random((2,3))  # Generating a 2D array of random numbers with shape (2,3)
print(b)  # Printing the array of random numbers

[[0.51182162 0.9504637  0.14415961]
 [0.94864945 0.31183145 0.42332645]]


In [43]:
print(b)
b+=a
print(b)

[[0.51182162 0.9504637  0.14415961]
 [0.94864945 0.31183145 0.42332645]]
[[3.51182162 3.9504637  3.14415961]
 [3.94864945 3.31183145 3.42332645]]


In [44]:
print(np.random.rand(2,5))
print(np.random.randint(0,10,(3,4)))  # Generating a 2D array of random integers between 0 and 10 with shape (3,4)


[[0.73813122 0.43388178 0.5936374  0.52562809 0.70030147]
 [0.13193732 0.18230137 0.46690961 0.53305976 0.55834026]]
[[8 7 4 6]
 [0 6 1 5]
 [4 3 9 1]]


In [45]:
a=np.random.randint(0,10,(3,4))  # Generating a 2D array of random integers between 0 and 10 with shape (3,4)
print(a)  # Printing the generated array

[[7 4 2 6]
 [1 2 5 6]
 [0 3 8 3]]


## 9. Universal Functions (ufuncs) & Aggregations
ufuncs like `np.exp`, `np.sqrt`, `np.sin` operate element-wise on arrays. Aggregation functions like `np.min`, `np.max`, `np.sum`, `np.cumsum` reduce an array to summary statistics.

In [46]:
print(np.exp(a))
print(np.sqrt(a))
print(np.sin(a))
print(np.min(a))
print(np.max(a))

[[1.09663316e+03 5.45981500e+01 7.38905610e+00 4.03428793e+02]
 [2.71828183e+00 7.38905610e+00 1.48413159e+02 4.03428793e+02]
 [1.00000000e+00 2.00855369e+01 2.98095799e+03 2.00855369e+01]]
[[2.64575131 2.         1.41421356 2.44948974]
 [1.         1.41421356 2.23606798 2.44948974]
 [0.         1.73205081 2.82842712 1.73205081]]
[[ 0.6569866  -0.7568025   0.90929743 -0.2794155 ]
 [ 0.84147098  0.90929743 -0.95892427 -0.2794155 ]
 [ 0.          0.14112001  0.98935825  0.14112001]]
0
8


same way sum cummsum sort avg etc ..

## 10. Indexing, Slicing & Iterating
NumPy supports Python-style slicing (`a[2:]`) extended to multiple axes (`a[2:, 2:, 1:]`). Iterating over a multi-dimensional array iterates over its first axis; use `.flat` to iterate over every scalar element regardless of dimensionality.

***Indexing slicing iterating***

In [47]:
a=np.array([[[1,2,3],[4,5,6],[7,8,9]],[[10,11,12],[13,14,15],[16,17,18]],[[19,20,21],[22,23,24],[25,26,27]]])  # Creating a 3D array with shape (3,3,3)
print(a)  # Printing the 3D array 'a'
print(a.shape)

[[[ 1  2  3]
  [ 4  5  6]
  [ 7  8  9]]

 [[10 11 12]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]
(3, 3, 3)


In [48]:
print(a[0,0])
print(a)
a[0,0]=[2,6,8]
print("after modification")
print(a)  # Printing the modified 3D array 'a'


[1 2 3]
[[[ 1  2  3]
  [ 4  5  6]
  [ 7  8  9]]

 [[10 11 12]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]
after modification
[[[ 2  6  8]
  [ 4  5  6]
  [ 7  8  9]]

 [[10 11 12]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]


In [49]:
print(a[2:])

[[[19 20 21]
  [22 23 24]
  [25 26 27]]]


In [50]:
print(a[2:,2:])

[[[25 26 27]]]


In [51]:
print(a[2:,2:,1:])

[[[26 27]]]


In [52]:
for x in a:
    print(x)  # Printing each 2D slice of the 3D array 'a'
    print(".......")
    for y in x:
        print(y)  # Printing each 1D slice of the 2D array 'x'
        print(".......")
        for z in y:
            print(z)  # Printing each element of the 1D array 'y'

[[2 6 8]
 [4 5 6]
 [7 8 9]]
.......
[2 6 8]
.......
2
6
8
[4 5 6]
.......
4
5
6
[7 8 9]
.......
7
8
9
[[10 11 12]
 [13 14 15]
 [16 17 18]]
.......
[10 11 12]
.......
10
11
12
[13 14 15]
.......
13
14
15
[16 17 18]
.......
16
17
18
[[19 20 21]
 [22 23 24]
 [25 26 27]]
.......
[19 20 21]
.......
19
20
21
[22 23 24]
.......
22
23
24
[25 26 27]
.......
25
26
27


In [53]:
for element in a.flat:
    print(element)  # Printing each element of the 3D array 'a' in a flattened manner

2
6
8
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27


## 11. `flatten()` vs `ravel()`
Both collapse an array to 1D, but **`flatten()` always returns a copy** (independent memory), while **`ravel()` returns a view whenever possible** (shares memory with the original). Modifying a raveled array can therefore modify the original array too.

In [54]:
b=a
print(b)
print(a.flatten())
print(a)
print(b)
print(a.ravel())
print(a)
print(b)

[[[ 2  6  8]
  [ 4  5  6]
  [ 7  8  9]]

 [[10 11 12]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]
[ 2  6  8  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27]
[[[ 2  6  8]
  [ 4  5  6]
  [ 7  8  9]]

 [[10 11 12]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]
[[[ 2  6  8]
  [ 4  5  6]
  [ 7  8  9]]

 [[10 11 12]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]
[ 2  6  8  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27]
[[[ 2  6  8]
  [ 4  5  6]
  [ 7  8  9]]

 [[10 11 12]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]
[[[ 2  6  8]
  [ 4  5  6]
  [ 7  8  9]]

 [[10 11 12]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]


In [55]:

arr = np.array([[1, 2],
                [3, 4]])

In [56]:
r = arr.ravel()
r[0] = 100

print(r)
print(arr)

[100   2   3   4]
[[100   2]
 [  3   4]]


In [57]:
arr = np.array([[1, 2],
                [3, 4]])

f = arr.flatten()
f[0] = 100

print(f)
print(arr)

[100   2   3   4]
[[1 2]
 [3 4]]


ravel()-- memory is shared so it affects the original array
flatten()--return new array

## 12. Reshaping Arrays
`reshape()` returns a new view/copy with a different shape (element count must match). `resize()` modifies the array **in place**, and can pad with zeros or truncate if the new size doesn't match the original.

In [58]:
a=np.random.randint(0,10,(3,4)) 
print(a)

[[1 8 9 8]
 [9 8 7 9]
 [3 0 3 4]]


In [59]:
a.shape

(3, 4)

**Reshaping the array**

In [60]:
a.reshape(2,6)

array([[1, 8, 9, 8, 9, 8],
       [7, 9, 3, 0, 3, 4]], dtype=int32)

In [61]:
a.resize(2, 6)
a

array([[1, 8, 9, 8, 9, 8],
       [7, 9, 3, 0, 3, 4]], dtype=int32)

In [62]:
a.reshape(6,-1)  # Reshaping the array 'a' to have 6 rows and automatically determine the number of columns

array([[1, 8],
       [9, 8],
       [9, 8],
       [7, 9],
       [3, 0],
       [3, 4]], dtype=int32)

## 13. Stacking Arrays
`np.vstack()` stacks arrays row-wise (axis 0), `np.hstack()` stacks column-wise (axis 1), and `np.column_stack()` is like `hstack` but treats 1D arrays as columns.

Stacking together diff ways
vstack(vertical stack) hstack (horizontal stack)

In [63]:
## to generate random numbers best practices is:
rg=np.random.default_rng(1)  # Creating a random number generator with a seed of 1

In [64]:
a=np.floor(10*rg.random((3,4)))  # Generating a 2D array of random numbers between 0 and 10, and applying the floor function to round down to the nearest integer
b=np.floor(10*rg.random((3,4)))  # Generating another 2D array of random numbers between 0 and 10, and applying the floor function to round down to the nearest integer

In [65]:
print(a)
print(b)

[[5. 9. 1. 9.]
 [3. 4. 8. 4.]
 [5. 0. 7. 5.]]
[[3. 7. 3. 4.]
 [1. 4. 2. 2.]
 [7. 2. 4. 9.]]


In [66]:
c=np.vstack((a, b))
print(c)
print(c.shape)

[[5. 9. 1. 9.]
 [3. 4. 8. 4.]
 [5. 0. 7. 5.]
 [3. 7. 3. 4.]
 [1. 4. 2. 2.]
 [7. 2. 4. 9.]]
(6, 4)


In [67]:
c=np.hstack((a, b))
print(c)
print(c.shape)

[[5. 9. 1. 9. 3. 7. 3. 4.]
 [3. 4. 8. 4. 1. 4. 2. 2.]
 [5. 0. 7. 5. 7. 2. 4. 9.]]
(3, 8)


In [68]:
c=np.column_stack((a, b))
print(c) ## same like hstack but it is for 2d array

[[5. 9. 1. 9. 3. 7. 3. 4.]
 [3. 4. 8. 4. 1. 4. 2. 2.]
 [5. 0. 7. 5. 7. 2. 4. 9.]]


In [69]:
a=np.array([4,2])
b=np.array([3,8])
np.column_stack((a, b))  # Stacking 1D arrays as columns into a 2D array

array([[4, 3],
       [2, 8]])

## 14. Splitting Arrays
`np.hsplit(a, n)` splits an array into `n` equal parts along columns; passing a tuple of indices instead splits **after** those specific column positions.

**Splittimg one array into several smaller ones**

In [70]:
a=np.floor(10*rg.random((2,12)))
print(a)

[[9. 7. 5. 2. 1. 9. 5. 1. 6. 7. 6. 9.]
 [0. 5. 4. 0. 6. 8. 5. 2. 8. 5. 5. 7.]]


In [71]:
np.hsplit(a,3)

[array([[9., 7., 5., 2.],
        [0., 5., 4., 0.]]),
 array([[1., 9., 5., 1.],
        [6., 8., 5., 2.]]),
 array([[6., 7., 6., 9.],
        [8., 5., 5., 7.]])]

In [72]:
np.hsplit(a,(3,4)) # Split `a` after the third and the fourth column

[array([[9., 7., 5.],
        [0., 5., 4.]]),
 array([[2.],
        [0.]]),
 array([[1., 9., 5., 1., 6., 7., 6., 9.],
        [6., 8., 5., 2., 8., 5., 5., 7.]])]

In [73]:
np.hsplit(a,(3,4,7,8))

[array([[9., 7., 5.],
        [0., 5., 4.]]),
 array([[2.],
        [0.]]),
 array([[1., 9., 5.],
        [6., 8., 5.]]),
 array([[1.],
        [2.]]),
 array([[6., 7., 6., 9.],
        [8., 5., 5., 7.]])]

## 15. Copies vs Views vs Assignment
Three very different behaviors when "copying" an array:

| Method | New object? | Shares data? | Changes propagate? |
|---|---|---|---|
| `b = a` (assignment) | No (same object) | Yes | Yes, both ways |
| `b = a.view()` | Yes | Yes (same buffer) | Yes, both ways |
| `b = a.copy()` | Yes | No (new buffer) | No, fully independent |


**Copies and views**

In [74]:
a = np.array([[ 0,  1,  2,  3],
              [ 4,  5,  6,  7],
              [ 8,  9, 10, 11]])

In [75]:
b=a
print(b)
print("......")
print(a)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
......
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


In [76]:
b[1,0]=3

In [77]:
print(b)

[[ 0  1  2  3]
 [ 3  5  6  7]
 [ 8  9 10 11]]


In [78]:
print(a)

[[ 0  1  2  3]
 [ 3  5  6  7]
 [ 8  9 10 11]]


In [79]:
## both a and b shares same memory 

In [80]:
a = np.array([[ 0,  1,  2,  3],
              [ 4,  5,  6,  7],
              [ 8,  9, 10, 11]])

In [81]:
b=a.view()
print(b)
print("......")
print(a)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
......
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


In [82]:
b[1,0]=3
print(a)
print("....")
print(b)

[[ 0  1  2  3]
 [ 3  5  6  7]
 [ 8  9 10 11]]
....
[[ 0  1  2  3]
 [ 3  5  6  7]
 [ 8  9 10 11]]


In [83]:
a = np.array([[ 0,  1,  2,  3],
              [ 4,  5,  6,  7],
              [ 8,  9, 10, 11]])

In [84]:
b=a.copy()
print(b)
print("......")
print(a)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
......
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


In [85]:
b[1,0]=3
print(a)
print("....")
print(b)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
....
[[ 0  1  2  3]
 [ 3  5  6  7]
 [ 8  9 10 11]]


by creating a var by view just refer to the memory location  so any modification can affect original array.

by using copy new data is created i.e In new memory location the data is stored and refered there.

## 16. Sorting Arrays
`np.sort(a)` returns a sorted copy — for a 2D array, each row is sorted independently by default.

In [86]:
a=np.floor(100*rg.random((4,5)))
print(a)

[[14. 81. 68. 78. 19.]
 [80. 19.  8. 85. 86.]
 [87. 47. 27.  0. 64.]
 [71. 83. 28. 21. 63.]]


In [87]:
a=np.sort(a) ## each row is sorted
print(a)

[[14. 19. 68. 78. 81.]
 [ 8. 19. 80. 85. 86.]
 [ 0. 27. 47. 64. 87.]
 [21. 28. 63. 71. 83.]]


## 17. Concatenation
`np.concatenate((a, b), axis=...)` joins arrays along an existing axis. `axis=0` joins rows (like vstack), `axis=1` joins columns (like hstack).

In [88]:
a = np.array([[1, 2, 3, 4],[5,6,7,8]])
b = np.array([[9,10,11,12],[13,14,15,16]])

In [89]:
np.concatenate((a,b))

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [90]:
np.concatenate((a,b),axis=0)

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [91]:
np.concatenate((a,b),axis=1)

array([[ 1,  2,  3,  4,  9, 10, 11, 12],
       [ 5,  6,  7,  8, 13, 14, 15, 16]])

In [92]:
a = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])

## 18. Boolean Indexing & Masking
Comparison operators create a boolean mask (`a < 5`) that can be used directly to select elements. `np.nonzero()` returns the indices where a condition is true instead of the values.

In [93]:
print(a[a<5])
print(a[a%2==0])
print(np.nonzero(a<5)) #to print the indices of elements that are less than 5:

[1 2 3 4]
[ 2  4  6  8 10 12]
(array([0, 0, 0, 0]), array([0, 1, 2, 3]))


## 19. Broadcasting
Broadcasting lets NumPy perform arithmetic between arrays of different (but compatible) shapes without explicitly copying data — smaller arrays are conceptually "stretched" to match the larger one.

In [94]:
data = np.array([1.0, 2.0])
data * 1.6 ## this is called broadcasting operation btw scalar and vector

array([1.6, 3.2])

## 20. Axis-wise Aggregation
Most aggregation functions accept an `axis` argument: `axis=0` aggregates down each column, `axis=1` aggregates across each row.

In [95]:
data = np.array([[1, 2], [5, 3], [4, 6]])
data
print(data.max(axis=0)) ## max at each column in col1 --5 col2 -- 6
print(data.max(axis=1)) ## max at each row

[5 6]
[2 5 6]


## 21. Unique Elements
`np.unique()` returns the sorted unique elements of an array. With `axis=0` on a 2D array it returns unique **rows**; `return_index`, `return_inverse`, and `return_counts` provide extra metadata about where those unique values came from.

In [96]:
a = np.array([11, 11, 12, 13, 14, 15, 16, 17, 12, 13, 11, 14, 18, 19, 20])
np.unique(a)

array([11, 12, 13, 14, 15, 16, 17, 18, 19, 20])

In [97]:
a_2d = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [1, 2, 3, 4]])
unique_values = np.unique(a_2d)
print(unique_values)

[ 1  2  3  4  5  6  7  8  9 10 11 12]


In [98]:
unique_rows = np.unique(a_2d, axis=0)
print(unique_rows)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


To get the unique rows, index position, and occurrence count

In [99]:
unique_rows,indexes,occ_count=np.unique(a_2d,axis=0,return_counts=True,return_index=True)
print(unique_rows)
print(indexes)
print(occ_count)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]
[0 1 2]
[2 1 1]


## 22. Transpose
For a 2D array of shape `(x, y)`, `a.T` (or `a.transpose()`) returns a view with shape `(y, x)` — rows and columns are swapped.

In [100]:
## Transpose matrix 
## a.T or a.transpose() if shape of a is (x,y) then a.T shape is (y,x)


In [101]:
arr_2d = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])

## 23. Flipping Arrays
`np.flip(a)` reverses element order; use `axis=0` to flip rows (upside-down) or `axis=1` to flip columns (left-right).

In [102]:
np.flip(arr_2d)

array([[12, 11, 10,  9],
       [ 8,  7,  6,  5],
       [ 4,  3,  2,  1]])

In [103]:
np.flip(arr_2d,axis=0)

array([[ 9, 10, 11, 12],
       [ 5,  6,  7,  8],
       [ 1,  2,  3,  4]])

In [104]:
np.flip(arr_2d,axis=1)

array([[ 4,  3,  2,  1],
       [ 8,  7,  6,  5],
       [12, 11, 10,  9]])

**if you see for occurence count , indexes we don't know syntax we refer from doc generally** but without that we can do

## 24. Using `help()` for Documentation
When you forget a function's exact signature or optional arguments, `help(np.function_name)` prints its full docstring — including parameters, return values, and usage examples — right in the notebook.

In [105]:
help(np.unique) ## docstring of that function or module

Help on _ArrayFunctionDispatcher in module numpy:

unique(ar, return_index=False, return_inverse=False, return_counts=False, axis=None, *, equal_nan=True, sorted=True)
    Find the unique elements of an array.

    Returns the sorted unique elements of an array. There are three optional
    outputs in addition to the unique elements:

    * the indices of the input array that give the unique values
    * the indices of the unique array that reconstruct the input array
    * the number of times each unique value comes up in the input array

    Parameters
    ----------
    ar : array_like
        Input array. Unless `axis` is specified, this will be flattened if it
        is not already 1-D.
    return_index : bool, optional
        If True, also return the indices of `ar` (along the specified axis,
        if provided, or in the flattened array) that result in the unique array.
    return_inverse : bool, optional
        If True, also return the indices of the unique array (for the s

## 25. Linear Algebra (Inverse & Determinant)
`numpy.linalg.inv(A)` computes the inverse of a square matrix A, such that `A @ A_inv` gives (approximately) the identity matrix. `numpy.linalg.det(A)` computes its determinant.

In [106]:
import numpy as np

# Define the matrix
A = np.array([[4, 7], [2, 6]])

# Compute the matrix inverse
A_inv = np.linalg.inv(A)
print(A_inv)

[[ 0.6 -0.7]
 [-0.2  0.4]]


In [107]:
import numpy as np

# Define the matrix
A = np.array([[1, 2], [3, 4]])

# Calculate the determinant
det_A = np.linalg.det(A)
print("Determinant of A:", det_A)

Determinant of A: -2.0000000000000004


## 26. Masked Arrays

A masked array in NumPy is a special type of array that includes an additional Boolean mask, which marks certain elements as invalid or masked. This allows you to work with data that has missing or invalid values without having to modify the original data.

In [108]:
import numpy as np
import numpy.ma as ma

# Create a normal NumPy array
data = np.array([1, 2, -999, 4, 5])

# Mask invalid values (-999 treated as missing data)
masked_data = ma.masked_equal(data, -999)

print("Original Data:", data)
print("Masked Data:", masked_data)

# Perform operations while ignoring masked values
mean_value = masked_data.mean()
print("Mean (ignoring masked values):", mean_value)

Original Data: [   1    2 -999    4    5]
Masked Data: [1 2 -- 4 5]
Mean (ignoring masked values): 3.0


## 27. Interview Questions — Quick Revision (Appendix)

A rapid-fire recap of the concepts most often asked about NumPy in interviews. Each point links back to the sections above — use this as a final revision pass.


**Q1. What is NumPy?**

An open-source Python library for fast numerical computing, built around the homogeneous, multi-dimensional `ndarray` and a large collection of vectorized mathematical routines. Arrays are created with `np.array()`, `np.zeros()`, `np.ones()`, `np.full()`, `np.arange()`, or `np.linspace()` (Sections 3, 5, 6).

**Q2. Main features of NumPy**

Speed and memory efficiency from contiguous storage, a large library of mathematical functions, broadcasting, easy integration with pandas/SciPy/scikit-learn, native multi-dimensional arrays, rich indexing/slicing, and fine control over memory layout (dtype, strides).

**Q3. Dot product of two arrays**

Use `a.dot(b)`, `np.dot(a, b)`, or the `@` operator — all three compute the same matrix/vector product (Section 7).

**Q4. Shallow copy vs deep copy**

A shallow copy (`a.view()`, or plain assignment `b = a`) shares the same underlying data buffer, so changes in one are visible in the other. A deep copy (`a.copy()`) allocates independent memory, so the two arrays never affect each other (Section 15).

**Q5. Reshaping an array**

`array.reshape(new_shape)` or `np.reshape(array, new_shape)` both return a new view/copy with the given shape, as long as the total element count is unchanged (Section 12).

**Q6. Element-wise operations**

Standard operators (`+ - * /`) and `np.power()` apply automatically element-by-element when arrays share the same shape — no explicit loop required (Section 7).

**Q7. Generating random numbers**

`np.random.rand()` gives uniform floats in [0,1), `np.random.randint()` gives random integers in a range, `np.random.randn()` samples from a standard normal distribution, and `np.random.seed()` (or the modern `np.random.default_rng(seed)`) makes results reproducible (Section 8).

**Q8. Array from a Python list**

Pass the list straight into `np.array()`, e.g. `np.array([1, 2, 3, 4, 5])` (Section 3).

**Q9. Conditional (boolean) indexing**

A comparison like `arr > 3` produces a boolean array; indexing with that boolean array (`arr[arr > 3]`) returns only the elements where the condition is true (Section 18).

**Q10. Common NumPy dtypes**

int, float, complex, bool, object, and datetime, among others — dtype controls how each element is stored in memory and what operations are valid on it.

**Q11. Vertical concatenation**

`np.vstack((a, b))` or `np.concatenate((a, b), axis=0)` stack arrays row-wise (Sections 13, 17).

**Q12. Matrix inversion**

`numpy.linalg.inv(A)` returns the inverse of a square matrix A such that `A @ A_inv` is the identity matrix (Section 25).

**Q13. `var()` and `mean()`**

`np.var(a, axis=None)` computes variance (spread of the data); `np.mean(a, axis=None)` computes the arithmetic average. Both accept an `axis` to reduce along rows or columns instead of the whole array (Section 20).

**Q14. Flattening a multi-dimensional array**

`array.flatten()` returns an independent 1D copy; `array.ravel()` returns a 1D view that shares memory with the original whenever possible (Section 11).

**Q15. Identifying outliers**

Two common approaches: (a) flag values further than *k* standard deviations from the mean, or (b) use the IQR rule — flag values below `Q1 - 1.5*IQR` or above `Q3 + 1.5*IQR`, computed with `np.percentile()`.

**Q16. Removing NaN / missing values**

Build a boolean mask with `np.isnan(arr)` and index with its inverse (`arr[~mask]`), or use `np.ma.masked_invalid(arr).compressed()` to drop invalid entries via a masked array (Section 26).

**Q17. Slicing vs indexing**

Slicing (`arr[1:5]`) extracts a contiguous range using `:` and returns a view; indexing (`arr[2]`, `arr[[1,3,5]]`) picks out specific, possibly non-contiguous elements (Section 10).

**Q18. Creating an array of repeated values**

Use `np.full(shape, value)`, or broadcasting like `7 * np.ones(5)`, or a list comprehension turned into an array (Section 5).

**Q19. What is a masked array?**

A NumPy array paired with a boolean mask marking certain entries as invalid, so statistics like `.mean()` automatically skip the masked values without altering the original data (Section 26).

**Q20. Broadcasting**

NumPy's mechanism for combining arrays of different but compatible shapes without physically duplicating data — a smaller array (or scalar) is conceptually stretched to match the shape of the larger one during the operation (Section 19).

**Q21. Sorting ascending/descending**

`np.sort(arr)` sorts ascending and returns a new array. For descending order, sort ascending and reverse it (`np.sort(arr)[::-1]`), or use `arr[np.argsort(-arr)]` (Section 16).

**Q22. Why arrays beat plain Python lists**

Better performance and memory layout (contiguous, homogeneous storage), native support for vectorized math, broadcasting, and true multi-dimensional structure — all backed by compiled C code.

**Q23. `reshape()` vs `resize()`**

`reshape()` returns a new view/copy without touching the original (element count must match exactly). `resize()` modifies the array in place and can even pad with zeros or truncate data if the new size differs from the original (Section 12).

**Q24. `vstack()` vs `hstack()`**

`vstack()` stacks arrays along axis 0 (adds rows, requires matching column counts); `hstack()` stacks along axis 1 (adds columns, requires matching row counts) (Section 13).

**Q25. Eigenvalues and determinant**

`np.linalg.eigvals(matrix)` returns the eigenvalues of a matrix; `np.linalg.det(matrix)` returns its determinant (Section 25).

**Q26. Comparing two arrays**

`(arr1 == arr2).all()` checks element-wise equality and reduces it to a single boolean, while `np.array_equal(arr1, arr2)` checks both shape and values in one call.

**Q27. QR decomposition**

`numpy.linalg.qr(a, mode='reduced')` factors a matrix A into an orthogonal matrix Q and an upper-triangular matrix R such that A = Q @ R.

**Q28. What is an ndarray?**

NumPy's core data structure: a homogeneous (single dtype), potentially N-dimensional array optimized for fast mathematical and array-oriented operations (Section 4).

**Q29. Vectorization**

Performing an operation on an entire array at once (e.g. `arr ** 2`) instead of looping element by element in Python — NumPy pushes the loop down into compiled C code, which is far faster (Section 7).

**Q30. `copy()` vs `view()` vs `=`**

`b = a` creates another reference to the *same* object (fastest, zero copy). `b = a.view()` creates a new array object with independent shape/attributes but shared data. `b = a.copy()` creates a fully independent array with its own data (slowest, but safest) (Section 15).

**Q31. `shape` vs `size`**

`shape` is a tuple describing the length along each axis (e.g. `(3, 4)`); `size` is a single integer — the total element count, i.e. the product of the shape tuple (Section 4).

**Q32. Python sequences vs NumPy arrays vs pandas**

Python lists/tuples can hold mixed types and are slow for numeric work. NumPy arrays are homogeneous, n-dimensional, and fast (C-backed). pandas Series/DataFrame are built on top of NumPy but add labeled indexing, alignment, and easier handling of missing data — better suited to tabular data analysis.

**Q33. DataFrame to NumPy array**

Use the `.values` attribute (or the more modern `.to_numpy()`) on a pandas DataFrame to get the underlying NumPy array.

**Q34. Reversing an array**

Slice with a step of -1: `arr[::-1]` (Section 10).

**Q35. Why NumPy is faster than lists**

Homogeneous, contiguous memory (better CPU cache use), vectorized C-level loops instead of Python-level loops, and use of optimized linear-algebra libraries (BLAS/LAPACK) under the hood.

**Q36. Counting occurrences of integers**

`np.bincount(arr)` counts how many times each non-negative integer appears — index i of the result is the count of value i in the array.

**Q37. Max / min of an array**

`np.max(arr)` and `np.min(arr)` (or `arr.max()`/`arr.min()`) return the largest/smallest element; pass `axis=` to reduce along rows or columns instead of the whole array (Section 20).

**Q38. Slicing/indexing for data cleaning**

Boolean indexing can replace values matching a condition in place (e.g. `data[data < 0] = 0` to clip negatives), while slicing/boolean selection can filter out a clean subset (e.g. `data[data > 2]`) (Section 18).

**Q39. Finding unique elements**

`np.unique(array)` returns the sorted array of distinct values, with optional flags (`return_index`, `return_counts`, `return_inverse`) for extra metadata (Section 21).